# Compare two Population Means
<center><img src="Images/Mean2/3.png"></center>

In [3]:
import pandas as pd
from pandas import DataFrame
import numpy as np
from scipy.stats import t, chi2, norm, f
from math import sqrt


## Inference about Two Population Means When Population Standard Deviations Are Known
<center><img src="Images/Mean2/4.png"></center>
<center><img src="Images/Mean2/1.png"></center>
<center><img src="Images/Mean2/9.png"></center>

## Example (Testing two population means are equal, σ known) 
Suppose two machines, say M1 and M2, are packaging 6-oz cans of talcum powder. It is known from the past behavior of the
machines that the weights of their respective fillings are normal with standard deviations of 0.04 oz and 0.05 oz, respectively. Suppose 100 cans filled by each machine are emptied, the contents are carefully weighed, and the sample averages are $\bar{X_1} = 6.11 oz$ and $\bar{X_2} = 6.14 oz.$ We wish to test at the α = 0.01 level of significance the hypothesis<br>
$H_0: μ_1 = μ_2$<br>
$H_1: μ_1 ≠ μ_2$<br>
Where μ1 and μ2 are means of populations of weights of fillings produced by machines M1 and M2, respectively.

In [8]:
alpha = 0.01
xbar1 = 6.11
xbar2 = 6.14

n1 = 100
n2 = 100

std1 = 0.04
std2 = 0.05

In [9]:
t_stat = (xbar1 - xbar2) / sqrt(std1**2 / n1 + std2**2 / n2)
t_stat

-4.685212856658082

In [11]:
zl = norm.ppf(alpha/2)
zr = norm.ppf(1-alpha/2)
print(f"Acceptance region is: ({zl}, {zr})")

Acceptance region is: (-2.575829303548901, 2.5758293035489004)


In [13]:
p_value = norm.cdf(t_stat)
p_value

1.3983456537314122e-06

**Interpretation:** Because t_stat lies out of the acceptance region, and p_value is less than 0.01 (significance level) we rejcet the null hypothesis in favor of the alternative hypothesis. So, the machines' means are not equal.

## Inference about Two Population Means When Population Standard Deviations Are Unknown and Equal

<center><img src="Images/Mean2/2.png"></center>
<center><img src="Images/Mean2/5.png"></center>
<center><img src="Images/Mean2/6.png"></center>


## Inference about Two Population Means When Population Standard Deviations Are Unknown and not Equal
<center><img src="Images/Mean2/7.png"></center>
<center><img src="Images/Mean2/8.png"></center>

If n1 and n2 are strictly greater than 30 we use the normal distribution instead of t distribution and:
<center><img src="Images/Mean2/10.png"></center>

#### Example 7: Direct and Broker-Purchased Mutual Funds
Millions of investors buy mutual funds choosing from thousands of possibilities. Some funds can be purchased
directly from banks or other financial institutions whereas others must be purchased through brokers, who charge a fee for this service. This raises the question, Can investors do better by buying mutual funds directly than by purchasing
mutual funds through brokers?
<br>To help answer this question, a group of researchers randomly sampled the annual returns from mutual funds that can be acquired directly and mutual funds that are bought through brokers and recorded the net annual returns, which are the returns on investment after deducting all relevant fees.
<br>Can we conclude at the 5% significance level that directly purchased mutual funds outperform mutual funds bought through brokers?
<br><br>**Solution:** To answer the question, we need to compare the population of returns from direct and
the returns from broker-bought mutual funds. This problem objective–data type combination tells us that the
parameter to be tested is the difference between two means, μ1 − μ2.<br>
$ H_0: (μ_1 − μ_2) = 0 $ <br>
$ H_1: (μ_1 − μ_2) > 0 $ <br>

To decide which of the t-tests of $ μ_1 − μ_2 $ to apply, we conduct the F-test of $ \frac{σ_1^ 2}{σ_2^ 2}. $ <br>
$H_0: \frac{σ_1^ 2}{σ_2^ 2}=1 $ <br>
$H_1: \frac{σ_1^ 2}{σ_2^ 2}≠1 $

In [14]:
df = pd.read_csv("Data\\funds.csv", index_col='i')
df.head(3)
direct_variance, broker_variance = np.var(df, ddof=1, axis=0)
f_statistics = direct_variance / broker_variance
print(f"f statistics = {f_statistics}")

left_rejection = f.ppf(.025,49,49)
right_rejection = f.ppf(.975,49,49) # = 1/left
print(f"Left Rejection = {round(left_rejection,2)}")
print(f"Right Rejection = {round(right_rejection,2)}")
print(f"f statistics = {f_statistics}")
print(f"""\nif f statistics lies out of the ({round(left_rejection,2)},{round(right_rejection,2)}) interval, we reject the null hypothesis in favor of the alternative hypothesis and 
conclude that the variance of the two pupulations are different from each other""")

if f_statistics > left_rejection and f_statistics < right_rejection:
    print("\nPopulation variances are equal.")
else:
    print("\nPopulation variances are not equal.")


f statistics = 0.8649931121915829
Left Rejection = 0.57
Right Rejection = 1.76
f statistics = 0.8649931121915829

if f statistics lies out of the (0.57,1.76) interval, we reject the null hypothesis in favor of the alternative hypothesis and 
conclude that the variance of the two pupulations are different from each other

Population variances are equal.


There is not enough evidence to infer that the population variances differ. It follows that
we must apply the equal-variances t-test of $ μ_1 − μ_2$.

In [15]:
direct_variance, broker_variance = np.var(df, ddof=1, axis=0)
direct_mean, broker_mean = np.mean(df, axis=0)
print(f"Direct and Broker Variances: {direct_variance} {broker_variance}")
print(f"Direct and Broker Means: {direct_mean} {broker_mean}")


Direct and Broker Variances: 37.48817812244897 43.33927934693878
Direct and Broker Means: 6.631199999999999 3.7232


In [16]:
N = df.shape[0]
pooled_variance = ((N - 1) * direct_variance + ( N - 1) * broker_variance) / (N + N - 2)
pooled_variance

40.413728734693876

In [17]:
significance = 0.95
dof = N + N - 2
left_rejection = t.ppf(significance, df = dof)
print(f"The rejection region is: t > {left_rejection}")

The rejection region is: t > 1.6605512170440568


In [18]:
mu1 = mu2 = 1234
t_statistics = ((direct_mean - broker_mean ) - (mu1 - mu2)) / sqrt(pooled_variance * ( 1/N + 1/N))


In [19]:
print(f"t statistics = {t_statistics}")
print(f"The rejection region is: t > {left_rejection}")

if t_statistics > left_rejection:
    print("\nThe null hypothesis is rejected and the the population means are not equal.")
else:
    print("\nWe can not reject the null hypothesis so the population means are equal.")

t statistics = 2.2871778974657735
The rejection region is: t > 1.6605512170440568

The null hypothesis is rejected and the the population means are not equal.


In [20]:
p_value = t.sf(2.28, df = 98)
p_value

0.0123874630366776

The value of the test statistic is 2.29. The one-tail p-value is .0122. We observe that
the p-value of the test is small (and the test statistic falls into the rejection region). As
a result, we conclude that there is sufficient evidence to infer that on average directly
purchased mutual funds outperform broker-purchased mutual funds.

**Estimating $μ_1 − μ_2$: Equal-Variances**<br>
In addition to testing a value of the difference between two population means, we can also
estimate the difference between means. Next we compute the 95% confidence interval
estimate of the difference between the mean return for direct and broker mutual funds.

**Solution:** The confidence interval estimator of the difference between two means with equal
population variances is
<left><img src="Images/Mean2/11.png"></left>


In [ ]:
significance = 0.95
dof = N + N - 2
alpha = (1 - significance) / 2
t_alpha = t.ppf(alpha, df = dof)
direct_mean, broker_mean = np.mean(df, axis=0)
pooled_variance = ((N - 1) * direct_variance + ( N - 1) * broker_variance) / (N + N - 2)

LCL = (direct_mean - broker_mean) + t_alpha * sqrt(pooled_variance * (1/N +1/N))
UCL = (direct_mean - broker_mean) - t_alpha * sqrt(pooled_variance * (1/N +1/N))

print(f"The confidence interval estimator of μ1 − μ2 is: ({round(LCL,2)}, {round(UCL,2)})")

We estimate that the return on directly purchased mutual funds is on average between .39
and 5.43 percentage points larger than broker-purchased mutual funds.

#### Example 8: Effect of New CEO in Family-run Businesses
What happens to the family-run business when the boss’s son or daughter takes over?
Does the business do better after the change if the new boss is the offspring of the owner,
or does the business do better when an outsider is made chief executive officer (CEO)?
In pursuit of an answer, researchers randomly selected 140 firms between 1994
and 2002, 30% of which passed ownership to an offspring and 70% of which appointed
an outsider as CEO. For each company, the researchers calculated the operating
income as a proportion of assets in the year before and the year after the new CEO took
over. The change (operating income after—operating income before) in this variable was recorded and is listed next.
<br>Do these data allow us to infer that the effect of making
an offspring CEO is different from the effect of hiring an outsider as CEO?<br><br>
**Solution:** The objective is to compare two populations, and the data are interval. It follows that the
parameter of interest is the difference between two population means $μ_1 − μ_2$ , where $μ_1$
is the mean difference for companies where the owner’s son or daughter became CEO
and $μ_2$ is the mean difference for companies who appointed an outsider as CEO.
<br>To determine whether to apply the equal or unequal variances t-test, we use the
F-test of two variances.<br>
$H_0: \frac{σ_1^ 2}{σ_2^ 2}=1 $ <br>
$H_1: \frac{σ_1^ 2}{σ_2^ 2}≠1 $


In [ ]:
df = pd.read_csv("Data\\CEO.csv", index_col='i')
df.head(3)
offspring_variance, outsider_variance = np.var(df, ddof=1, axis=0)
f_statistics = offspring_variance / outsider_variance
print(f"f statistics = {f_statistics}")

N1 = df['Offspring'].count()
N2 = df['Outsider'].count()
nu1 = N1 - 1
nu2 = N2 - 2

left_rejection = f.ppf(.025,nu1, nu2)
right_rejection = f.ppf(.975,nu1, nu2) # = 1/left
print(f"Left Rejection = {round(left_rejection,2)}")
print(f"Right Rejection = {round(right_rejection,2)}")
print(f"f statistics = {f_statistics}")
print(f"""\nIf f statistics lies out of the ({round(left_rejection,2)},{round(right_rejection,2)}) interval, we reject the null hypothesis in favor of the alternative hypothesis and 
conclude that the variance of the two pupulations are different from each other""")
if f_statistics > left_rejection and f_statistics < right_rejection:
    print("\nPopulation variances are equal.")
else:
    print("\nPopulation variances are not equal.")

Because F = .47 is less than .58 , we reject the null hypothesis.<br><br>
There is enough evidence to infer that the population variances differ. The appropriate
technique is the unequal-variances t-test of $μ_1 − μ_2$.<br>
Because we want to determine whether there is a difference between means, the null and 
alternative hypothesises are:<br>

$ H_0: (μ_1 − μ_2) = 0 $ <br>
$ H_1: (μ_1 − μ_2) ≠ 0 $ <br>


In [ ]:
offspring_variance, outsider_variance = np.var(df, ddof=1, axis=0)
offspring_mean, outsider_mean = np.mean(df, axis=0)
print(f"Offspring and Outsider Variances: {offspring_variance} {outsider_variance}")
print(f"Offspring and Outsider Means: {offspring_mean} {outsider_mean}")

The number of degrees of freedom of the test statistic is:

<left><img src="Images/Mean2/12.png" width=250></left>

In [ ]:
nu = ((offspring_variance/N1 + outsider_variance/N2)**2) / ((offspring_variance/N1)**2 / nu1 + 
                                                    (outsider_variance/N2)**2 / nu2)
nu = np.ceil(nu)

significance = 0.95
alpha = (1 - significance) / 2
left_rejection = t.ppf(alpha, df=nu)
right_rejection = -t.ppf(alpha, df=nu)

print(f"The rejection region is: t < {left_rejection} or t > {right_rejection}")
                                              

The value of the test statistic is computed next:

<left><img src="Images/Mean2/13.png" width=250></left>


In [ ]:
mu1 = mu2 = 1234
t_statistics = ((offspring_mean - outsider_mean ) - (mu1 - mu2)) / sqrt(offspring_variance/N1 + outsider_variance/N2)

print(f"t statistics = {t_statistics}")
print(f"The rejection region is: t < {left_rejection} or t > {right_rejection}")

t.ppf(0.0008,df=nu)
if t_statistics < left_rejection or t_statistics > right_rejection:
    print("\nThe null hypothesis is rejected and the the population means are not equal.")
else:
    print("\nWe can not reject the null hypothesis so the population means are equal.")

p_value = t.sf(-t_statistics, df = nu)
round(p_value,4)

The t-statistic is −3.22, and its p-value is .0017. Accordingly, we conclude there is
sufficient evidence to infer that the mean changes in operating income differ.

**Estimating $μ_1 − μ_2$: unequal-Variances**<br>
We can also draw inferences about the difference between the two population means by
calculating the confidence interval estimator. We use the unequal-variances confidence
interval estimator of $μ_1 − μ_2$ and a 95% confidence level.<br>
<left><img src="Images/Mean2/14.png" width=250></left>


In [ ]:
significance = 0.95
alpha = (1 - significance) / 2
t_alpha = t.ppf(alpha, df = nu)
offspring_variance, outsider_variance = np.var(df, ddof=1, axis=0)
offspring_mean, outsider_mean = np.mean(df, axis=0)

LCL = (offspring_mean - outsider_mean) + t_alpha * sqrt(offspring_variance/N1 + outsider_variance/N2)
UCL = (offspring_mean - outsider_mean) - t_alpha * sqrt(offspring_variance/N1 + outsider_variance/N2)

print(f"The confidence interval estimator of μ1 − μ2 is: ({round(LCL,2)}, {round(UCL,2)})")

We estimate that the mean change in operating incomes for outsiders exceeds the mean
change in the operating income for offspring by between .52 and 2.16 percentage points.

## Inference about the Difference between Two Means: Matched Pairs Experiment
In the previous sections, we studied the testing of hypotheses about the difference of two population means on the basis of two independent random samples, one sample from each population. Quite often, for various reasons, the experiments are designed in such a way that the data are collected in pairs, that is, two observations are taken on the same
subject or experimental unit. Consequently, these two observations are not independent.<br>
Data collected in this manner are usually known as paired data. Since the data collected have two observations on each subject, there is an implicit dependency between the two samples, one collected before an action and the other afterward. However, if we use the techniques of testing hypotheses discussed earlier, where we had independent samples, then the results may turn out to be inaccurate.<br>
We now consider the problem that arises when we perform an experiment on n subjects, producing the n pairs of random variables (Y11,Y12),(Y21,Y22), ...,(Yn1,Yn2), where (Yi1,Yi2) is the pair of measurements obtained from the ith subject. It is assumed that the differences Yi2 − Yi1 = Xi, i = 1, ...,n, are n independent random variables all having identical normal distributions, namely N(μ,σ2). Thus, the n differences (X1, ...,Xn) are essentially a random sample of size n from N(μ,σ2). Let Xbar and S2 be the average and variance of this sample of X’s. Then, the problem is to test the hypothesis (μ = E(X) = E(Y1 − Y2)).<br>
$H_0: μ = μ_0$<br>
$H_1: μ ≠ μ_0$<br>

<center><img src="Images/Mean2/15.png" width=700></center>

<center><img src="Images/Mean2/16.png" width=700></center>
